File: src/main/javabak/tile_entities/capabilities/MergedIoFluidCapability.java
package redcrafter07.processed.block.tile_entities.capabilities;

import net.neoforged.neoforge.fluids.FluidStack;
import net.neoforged.neoforge.fluids.capability.IFluidHandler;
import net.neoforged.neoforge.fluids.capability.templates.EmptyFluidHandler;
import redcrafter07.processed.block.machine_abstractions.IoState;
import redcrafter07.processed.block.machine_abstractions.ProcessedMachine;

public class MergedIoFluidCapability implements IFluidHandlerModifiable {

    private final ProcessedMachine.CapabilityHandlers handlers;

    public MergedIoFluidCapability(ProcessedMachine.CapabilityHandlers handlers) {
        this.handlers = handlers;
    }

    private IFluidHandler get(IoState state) {
        final var handler = handlers.getFluidHandlerForState(state);
        if (handler == null) return EmptyFluidHandler.INSTANCE;
        return handler;
    }

    @Override
    public int getTanks() {
        return get(IoState.Output).getTanks() + get(IoState.Input).getTanks();
    }

    @Override
    public FluidStack getFluidInTank(int slot) {
        final var handler = handlers.getFluidHandlerForState(IoState.Output);
        if (handler == null) return FluidStack.EMPTY;
        if (slot >= handler.getTanks() || slot < 0) return FluidStack.EMPTY;
        return handler.getFluidInTank(slot);
    }

    @Override
    public void setFluidInTank(int slot, FluidStack stack) {
        var realSlot = slot - get(IoState.Output).getTanks();

        if (get(IoState.Input) instanceof IFluidHandlerModifiable handler) {
            if (realSlot < 0 || realSlot >= handler.getTanks()) return;
            handler.setFluidInTank(realSlot, stack);
        }
    }

    @Override
    public int fill(FluidStack resource, FluidAction action) {
        return get(IoState.Input).fill(resource, action);
    }

    @Override
    public FluidStack drain(int maxDrain, FluidAction action) {
        return get(IoState.Output).drain(maxDrain, action);
    }

    @Override
    public FluidStack drain(FluidStack resource, FluidAction action) {
        return get(IoState.Output).drain(resource, action);
    }

    @Override
    public int getTankCapacity(int slot) {
        final var handlerInput = get(IoState.Input);
        final var handlerOutput = get(IoState.Output);

        if (slot < 0) return 0;
        if (slot < handlerOutput.getTanks()) return handlerOutput.getTankCapacity(slot);
        if (slot < handlerInput.getTanks() + handlerOutput.getTanks()) return handlerInput.getTankCapacity(slot - handlerOutput.getTanks());
        return 0;
    }

    @Override
    public boolean isFluidValid(int slot, FluidStack stack) {
        final var handlerInput = get(IoState.Input);
        final var handlerOutput = get(IoState.Output);

        if (slot < 0) return false;
        if (slot < handlerOutput.getTanks()) return handlerOutput.isFluidValid(slot, stack);
        if (slot < handlerInput.getTanks() + handlerOutput.getTanks()) return handlerInput.isFluidValid(slot - handlerOutput.getTanks(), stack);
        return false;
    }
}

File: src/main/kotlin/redcrafter07/processed/block/tile_entities/capabilities/MergedIoFluidCapability.kt
package redcrafter07.processed.block.tile_entities.capabilities

import net.neoforged.neoforge.fluids.FluidStack
import net.neoforged.neoforge.fluids.capability.IFluidHandler
import net.neoforged.neoforge.fluids.capability.templates.EmptyFluidHandler
import redcrafter07.processed.block.machine_abstractions.IoState
import redcrafter07.processed.block.machine_abstractions.ProcessedMachine

class MergedIoFluidCapability(val handlers: ProcessedMachine.CapabilityHandlers): FluidHandlerModifiable {

    private fun get(state: IoState): IFluidHandler = handlers.getFluidHandlerForState(state) ?: EmptyFluidHandler.INSTANCE

    override fun getTanks(): Int {
        return get(IoState.Output).tanks + get(IoState.Input).tanks
    }

    override fun getFluidInTank(slot: Int): FluidStack {
        val handler = handlers.getFluidHandlerForState(IoState.Output) ?: return FluidStack.EMPTY
        if (slot >= handler.tanks || slot < 0) return FluidStack.EMPTY
        return handler.getFluidInTank(slot)
    }

    override fun setFluidInTank(tank: Int, stack: FluidStack) {
        val realSlot = tank - get(IoState.Output).tanks

        val handler = get(IoState.Input)
        if (handler is FluidHandlerModifiable) {
            if (realSlot < 0 || realSlot >= handler.tanks) return
            handler.setFluidInTank(realSlot, stack)
        }
    }

    override fun fill(resource: FluidStack, action: IFluidHandler.FluidAction): Int {
        return get(IoState.Input).fill(resource, action)
    }

    override fun drain(maxDrain: Int, action: IFluidHandler.FluidAction): FluidStack {
        return get(IoState.Output).drain(maxDrain, action)
    }

    override fun drain(resource: FluidStack, action: IFluidHandler.FluidAction): FluidStack {
        return get(IoState.Output).drain(resource, action)
    }

    override fun getTankCapacity(slot: Int): Int {
        val handlerInput = get(IoState.Input)
        val handlerOutput = get(IoState.Output)

        if (slot < 0) return 0
        if (slot < handlerOutput.tanks) return handlerOutput.getTankCapacity(slot)
        if (slot < handlerInput.tanks + handlerOutput.tanks) return handlerInput.getTankCapacity(slot - handlerOutput.tanks)
        return 0
    }

    override fun isFluidValid(slot: Int, stack: FluidStack): Boolean {
        val handlerInput = get(IoState.Input)
        val handlerOutput = get(IoState.Output)

        if (slot < 0) return false
        if (slot < handlerOutput.tanks) return handlerOutput.isFluidValid(slot, stack)
        if (slot < handlerInput.tanks + handlerOutput.tanks) return handlerInput.isFluidValid(
            slot - handlerOutput.tanks,
            stack
        )
        return false
    }
}